sway/commands/bind.c

```diff
@@ -649,101 +647,157 @@ void seat_execute_command(struct sway_seat *seat, struct sway_binding *binding)
+ * the specified keysym's list of matched keycodes.
  */
-static void find_keycode(struct xkb_keymap *keymap,
+static void add_matching_keycodes(struct xkb_keymap *keymap,
 		xkb_keycode_t keycode, void *data) {
 	xkb_keysym_t keysym = xkb_state_key_get_one_sym(
 			config->keysym_translation_state, keycode);
 
-	if (keysym == XKB_KEY_NoSymbol) {
+	struct keycode_matches *matches = data;
+	if (*matches->keysym == keysym) {
+		xkb_keycode_t *new_keycode = malloc(sizeof(xkb_keycode_t));
+		if (!new_keycode) {
+			matches->error = true;
+			return;
+		}
+		*new_keycode = keycode;
+		list_add(matches->keycodes, new_keycode);
+	}
+}
+
+void cartesian_product_helper(list_t **sets, int n, xkb_keycode_t** current_result, int* curr_size, xkb_keycode_t* current_product, int depth) {
+	// Conquer
+	if (depth == n) {
+		current_result[*curr_size] = malloc(n * sizeof(xkb_keycode_t));
+		for (int i = 0; i < n; ++i) {
+		current_result[*curr_size][i] = current_product[i];
+		}
+		(*curr_size)++;
 		return;
 	}
 
-	struct keycode_matches *matches = data;
-	if (matches->keysym == keysym) {
-		matches->keycode = keycode;
-		matches->count++;
+	// Divide
+	for (int i = 0; i < sets[depth]->length; ++i) {
+		current_product[depth] = *(xkb_keycode_t*)sets[depth]->items[i];
+		cartesian_product_helper(sets, n, current_result, curr_size, current_product, depth + 1);
 	}
 }
 
 /**
- * Return the keycode for the specified keysym.
+ * Compute the calculate the Cartesian product of `n` sets
  */
-static struct keycode_matches get_keycode_for_keysym(xkb_keysym_t keysym) {
-	struct keycode_matches matches = {
-		.keysym = keysym,
-		.keycode = XKB_KEYCODE_INVALID,
-		.count = 0,
-	};
-
-	xkb_keymap_key_for_each(
-			xkb_state_get_keymap(config->keysym_translation_state),
-			find_keycode, &matches);
-	return matches;
+xkb_keycode_t **cartesian_product(list_t **sets, int n) {
+	int total_combinations = 1;
+	for (int i = 0; i < n; ++i) {
+		total_combinations *= sets[i]->length;
+	}
+		
+	// Allocate memory for the current_result
+	int result_size = 0;
+	xkb_keycode_t **result = malloc(total_combinations * sizeof(xkb_keycode_t*));
+	xkb_keycode_t *current_product = malloc(n * sizeof(xkb_keycode_t));
+	cartesian_product_helper(sets, n, result, &result_size, current_product, 0);
+	free(current_product);
+
+	return result;
 }
 
+/*
+ * Convert keysyms to keycodes for --to-code. If any keysym does not have at
+ * least 1 keycode: halt conversion and unset BINDING_CODE. Assumes
+ * identify_key() read in binding->keys in keysym format vs keycode format
+ */
 bool translate_binding(struct sway_binding *binding) {
-	if ((binding->flags & BINDING_CODE) == 0) {
-		return true;
-	}
 
-	switch (binding->type) {
-	// a bindsym to translate
-	case BINDING_KEYSYM:
-		binding->syms = binding->keys;
-		binding->keys = create_list();
-		break;
-	// a bindsym to re-translate
-	case BINDING_KEYCODE:
-		list_free_items_and_destroy(binding->keys);
-		binding->keys = create_list();
-		break;
```

Comment:
For clarity of discussion, this is the retranslation logic that this (and my WIP) patch does not currently handle correctly. This serves to redo the `--to-code` conversion if the *primary* keymap changes, i.e., if you go from `us,ru` to `ru,us` or if you first set your non-standard keymap after all the bindyms. `--to-code` is not locked at the time of binding, but instead respect the configured primary keymap rather than the current keymap.

Reply:
I was already testing by switching between us,ru and it seemed to work in both cases. Like I would toggle the layout 

```config
### Input configuration
input 1:1:AT_Translated_Set_2_keyboard {
    xkb_layout us,ru
    xkb_options ctrl:swapcaps,grp:alt_shift_toggle
}
```

Which input did you test with and failed? So I can try on my machine too.

Reply:
```sh
swaymsg 'input * xkb_keymap us'
swaymsg 'bindsym --to-code mod1+a exec notify-send hi'
# test that alt+a sends hi according to a US QWERTY layout
swaymsg 'input * xkb_keymap fr'
# Test that alt+q now sends hi, as the French AZERTY layout is the new
# "first configured layout" that `--to-code` should respect
```

Reply:
That worked for me.

Reply:
> Test that alt+q now sends hi

Hold up I just saw this. That seems like a weird use case to me. If I bind something to code, I expect those keys to not change positions regardless of the layout.

Reply:
What's the purpose of even binding something to keycodes if you have to look up where the keycode is after changing layout? might as well bindsym.

Reply:
There’s a long discussion in the original PR, but the TL;DR is that it makes it so that the bindsym behavior is not dependent on command ordering - whether you set keyboard layout first or last in the config, the result is the same.

This makes it consistent with bindcode and regular bindsym, which also give the same result regardless of ordering in respect to keyboard configuration.

Reply:
I see, but that still completely defeats the purpose of binding things to keycodes. Isn't the purpose of bindcode so that when you change layouts, it will be the same physical keys on the keyboard?

Reply:
If bindcode and bindsym --to-code do not stick with the layout that my keyboard is configured in, that makes both commands obsolete. What you really want to do in that example is bindsym.

Reply:
`bindsym --to-code` is not the same as `bindcode`. Because it uses symbols and depends on a layout, it is implemented so that it makes the binding to match the keycodes of the at any time first configured layout.

Quoting the manpage:

```
	Bindings to keysyms are layout-dependent. This can be changed with the
	_--to-code_ flag. In this case, the keysyms will be translated into the
	corresponding keycodes in the first configured layout.
```

This means that when you have a keyboard configured for `xkb_layout us,fr`, a `bindsym --to-code` will  use the sequence according to the US keyboard layout even if you switch to French using `xkb_switch_layout` (`man 5 sway-input`) or an XKB-builtin layout switch toggle.

If the translation *only* happened once, then it would mean that the result depended on the order of `bindsym` and `xkb_layout` in your config file, which we do not want to be the case.

Reply:
> If bindcode and bindsym --to-code do not stick with the layout that my physical keyboard is configured in, that makes both commands obsolete. What you really want to do in that example is bindsym.

This is how the functionality is defined and documented. The PR and issue is just about adding support for symbols emitted by more than one keycode.

Reply:
> What's the "original PR" btw?

https://github.com/swaywm/sway/pull/3058

Reply:
> bindsym --to-code is not the same as bindcode

Thanks.

Reply:
> This is how the functionality is defined and documented

> Test that alt+q now sends hi

I should have tested on default sooner but I just realized that this behavior doesn't work with the current code as the documentation says. Can you verify?

If so, I can start looking into making it work as intended.